numba/cpython/builtins.py

```diff
@@ -673,3 +675,81 @@ def impl(func, iterable):
                 if func(x):
                     yield x
     return impl
+
+
+@overload(isinstance)
+def ol_isinstance(var, typs):
+    def true_impl(var, typs):
+        return True
+    def false_impl(var, typs):
+        return False
+
+    var_ty = as_numba_type(var)
+
+    if isinstance(var_ty, types.Optional):
+        raise NumbaTypeError(
+            f'isinstance cannot handle optional types. '
+            f'Got "{var_ty}"')
+
+    # register container types (List, Dict, ...) with types.Any as base type
+    as_numba_type.register(tuple, types.Tuple((types.undefined,)))
+    as_numba_type.register(list, types.List(types.undefined))
+    # set requires the type to be hashable
+    as_numba_type.register(set, types.Set(types.undefined))
+    as_numba_type.register(types.ListType, types.ListType(types.undefined))
+
+    t_typs = typs
+
+    if isinstance(t_typs, types.UniTuple):
+        # corner case - all types in isinstance are the same
+        t_typs = (t_typs.key[0])
+
+    if not isinstance(t_typs, types.Tuple):
+        t_typs = (t_typs, )
+
+    for typ in t_typs:
+        if isinstance(typ, types.Function):
+            key = typ.key[0]  # functions like int(..), float(..), str(..)
+        elif isinstance(typ, types.ClassType):
+            key = typ  # jitclasses
+        else:
+            key = typ.key
+
+        # corner cases for bytes, range, ...
+        # avoid registering those types on `as_numba_type`
+        types_not_registered = {
+            bytes: types.Bytes,
+            range: types.RangeType,
+            dict: (types.DictType, types.LiteralStrKeyDict),
+        }
+        if key in types_not_registered:
+            if isinstance(var_ty, types_not_registered[key]):
+                return true_impl
+            continue
+
+        if isinstance(typ, types.TypeRef):
+            # Case for TypeRef (i.e. isinstance(var, typed.List))
+            #      var_ty == ListType[int64] (instance)
+            #         typ == types.ListType  (class)
+            return true_impl if type(var_ty) is key else false_impl
+        elif isinstance(typ, types.NamedTupleClass):
+            # Case for NamedTuples
+            return true_impl if issubclass(var_ty.__class__, types.BaseNamedTuple) else false_impl
```

Comment:
I'm not sure that this is quite right because it's just checking in the typing domain? e.g. This breaks:
```python
from numba import njit
import collections

NT = collections.namedtuple('NT', 'x')
ZT = collections.namedtuple('ZT', 'x')

nt = NT(1)

@njit
def foo(a):
    return isinstance(a, NT), isinstance(a, ZT)

assert foo(nt) == foo.py_func(nt)

```

Reply:
Maybe something similar to this could help?:
```python
            if issubclass(var_ty.__class__, types.BaseNamedTuple) and\
                issubclass(var_ty.instance_class, typs.instance_class):
                return true_impl
            else:
                return false_impl
```